File: src/main/java/com/vsharstuk/pizzatoppings/controller/UserToppingController.java
package com.vsharstuk.pizzatoppings.controller;

import com.vsharstuk.pizzatoppings.dto.UserDto;
import com.vsharstuk.pizzatoppings.dto.UserToppingRequest;
import com.vsharstuk.pizzatoppings.entity.Topping;
import com.vsharstuk.pizzatoppings.entity.User;
import com.vsharstuk.pizzatoppings.mapper.ToppingMapper;
import com.vsharstuk.pizzatoppings.mapper.UserMapper;
import com.vsharstuk.pizzatoppings.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/users")
@RequiredArgsConstructor
public class UserToppingController {

    private final UserService userService;
    private final ToppingMapper toppingMapper;
    private final UserMapper userMapper;

    @PostMapping("/{user_id}/toppings")
    @ResponseStatus(HttpStatus.CREATED)
    public void submitToppings(@PathVariable("user_id") Long userIid,
                               @RequestBody @Valid UserToppingRequest request) {

        List<Topping> toppings = request.getToppings().stream().map(toppingMapper::toToppingEntity).collect(Collectors.toList());
        userService.submit(userIid, toppings);
    }

    @GetMapping("/{user_id}")
    public UserDto getUser(@PathVariable("user_id") Long userId) {
        return userMapper.toUserDto(userService.getUser(userId));
    }

    @PostMapping
    public UserDto create(@RequestBody @Valid UserDto userDto) {
        User user = userService.create(userMapper.toUserEntity(userDto));
        return userMapper.toUserDto(user);
    }
}

File: src/main/kotlin/com/vsharstuk/pizzatoppings/controller/UserToppingController.kt
package com.vsharstuk.pizzatoppings.controller

import com.vsharstuk.pizzatoppings.dto.ToppingDto
import com.vsharstuk.pizzatoppings.dto.UserDto
import com.vsharstuk.pizzatoppings.dto.UserToppingRequest
import com.vsharstuk.pizzatoppings.mapper.ToppingMapper
import com.vsharstuk.pizzatoppings.mapper.UserMapper
import com.vsharstuk.pizzatoppings.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import java.util.stream.Collectors
import javax.validation.Valid

@RestController
@RequestMapping("/users")
class UserToppingController {

    @Autowired
    lateinit var userService: UserService
    @Autowired
    lateinit var toppingMapper: ToppingMapper
    @Autowired
    lateinit var userMapper: UserMapper

    @PostMapping("/{user_id}/toppings")
    @ResponseStatus(HttpStatus.CREATED)
    fun submitToppings(@PathVariable("user_id") userIid: Long?,
                       @RequestBody request: @Valid UserToppingRequest?) {
        val toppings = request!!.toppings!!.stream().map { dto: ToppingDto? -> toppingMapper!!.toToppingEntity(dto) }.collect(Collectors.toList())
        userService!!.submit(userIid!!, toppings)
    }

    @GetMapping("/{user_id}")
    fun getUser(@PathVariable("user_id") userId: Long?): UserDto? {
        return userMapper!!.toUserDto(userService!!.getUser(userId!!))
    }

    @PostMapping
    fun create(@RequestBody userDto: @Valid UserDto?): UserDto? {
        val user = userService!!.create(userMapper!!.toUserEntity(userDto)!!)
        return userMapper.toUserDto(user)
    }
}
